File: packages/totem-sdk/packages/identity/src/verify.ts
/**
 * Claim verification.
 *
 * verifyIdentityClaim recomputes the canonical claim digest from signed.claim
 * using canonicalJson. It does NOT use proof.message as the source of truth —
 * that field is optional debug/display metadata only.
 *
 * The public-key-to-address binding is not checked here because the address
 * derivation from the public key requires internal state only available at
 * signing time. The WOTS signature itself binds the public key to the message;
 * callers who need an address-to-key binding should independently verify that
 * through an identity proof or other trust mechanism.
 */

import { wotsVerifyDigest, hexToBytes } from '@totemsdk/core';
import { claimDigest } from './signing.js';
import type { SignedIdentityClaim, IdentityVerifyResult } from './types.js';

export function verifyIdentityClaim(signed: SignedIdentityClaim): IdentityVerifyResult {
  const { claim, proof } = signed;

  let sigBytes: Uint8Array;
  let pkDigest: Uint8Array;
  try {
    sigBytes = hexToBytes(proof.signature);
    pkDigest = hexToBytes(proof.publicKey);
  } catch (e) {
    return {
      valid: false,
      reason: `hex decode failed: ${String(e)}`,
      signerAddress: proof.address,
    };
  }

  const digest = claimDigest(claim);

  let sigValid: boolean;
  try {
    sigValid = wotsVerifyDigest(sigBytes, digest, pkDigest);
  } catch (e) {
    return {
      valid: false,
      reason: `WOTS verify threw: ${String(e)}`,
      signerAddress: proof.address,
    };
  }

  if (!sigValid) {
    return {
      valid: false,
      reason: 'WOTS signature invalid',
      signerAddress: proof.address,
    };
  }

  return {
    valid: true,
    signerAddress: proof.address,
  };
}

File: packages/totem-sdk/packages/identity/rust/src/verify.rs
use crate::canonical::claim_digest;
use crate::types::{IdentityVerifyResult, SignedIdentityClaim};

pub fn verify_identity_claim(
    signed: &SignedIdentityClaim,
    verify_sig_fn: &dyn Fn(&str, &[u8], &str) -> bool,
) -> IdentityVerifyResult {
    let claim_json = serde_json::to_value(&signed.claim).unwrap_or_default();
    let digest = claim_digest(&claim_json);

    if !verify_sig_fn(&signed.proof.signature, &digest, &signed.proof.public_key) {
        return IdentityVerifyResult {
            valid: false,
            reason: Some("WOTS signature invalid".to_string()),
            signer_address: Some(signed.proof.address.clone()),
            root_address: None,
        };
    }

    IdentityVerifyResult {
        valid: true,
        reason: None,
        signer_address: Some(signed.proof.address.clone()),
        root_address: None,
    }
}
